Create constants for top-level rolenames

ivanayov opened the pull request:
Fixes #1648

This is a change in the metadata API to remove hardcoded role names
and use constants instead.

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature - updated only
- [ ] Docs have been added for the bug fix or new feature - n/a

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -61,14 +61,20 @@
     SignedSerializer,
 )
 
+class Rolename:
+    ROOT = "root"
+    TIMESTAMP = "timestamp"
+    SNAPSHOT = "snapshot"
+    TARGETS = "targets"
```
Generally, I would prefer if we use 4 separate constants like that:
```
ROOT = "root"
TIMESTAMP = "timestamp"
SNAPSHOT = "snapshot"
TARGETS = "targets"
```
instead of creating a separate class containing the constants.
The reason is that using directly the constants is significantly shorter also we don't expect to change the number of role names anytime soon and I am not sure what advantage do we get by using a class and then the constants nested inside.

ivanayov replied:
Thanks for the review @MVrachev !

There are 2 reasons I've put those in a class:
1. If a new role is being added in future
2. All "root", "timestamp", "snapshot", "targets" are role names in TUF's context, but excluding the specific use they are very general names and being explicitly referenced with a role name creates better readability.

On the other side I can agree that `Rolename` is used for nothing else than a constant "reference" and this is a bit overkill in terms of resources. 
So if 1. or 2. are not considered important, I can remove the class.

jku replied:
> 1. If a new role is being added in future

If a new role is added, python-tuf likely cannot cope with that without major code changes. So I don't think this structure needs to anticipate that either

MVrachev reviewed (changes requested):
Thanks for your contribution @ivanayov!
I have one preference about the way you defined your constants.
Still, I would want to hear an opinion from someone from the maintainers.
What do you think @sechkova?
Do you agree with my comment?

sechkova commented:
I agree with both of you ... 
On one hand, Root, Targets etc. classes already exist in metadata.py and now if we add similar constants for role names, we'll be importing things like 
```
from tuf.api.metadata import Targets, TARGETS
```
On the other hand, names do grow in size.

I was thinking if it there is a convenient data structure than would:
- encapsulate rolenames
- act as iterator allowing us to call list(Rolenames); for role in Rolenames: and by doing this replaces TOP_LEVEL_ROLE_NAMES too
- have a short enough name :shrug: 

But this is only a suggestion and if implementing this feels like an overkill for python-tuf, I'm ok with simple constants.

MVrachev commented:
@lukpueh do you have an opinion on the matter?
I am asking you because we had discussed such issues before with you.

lukpueh commented:
FWIW, [in warehouse they use an enum subclass](https://github.com/pypa/warehouse/blob/0524db370d1c2b5fd2d16433e0410d77a0b0c362/warehouse/tuf/constants.py#L16-L23 ) to do this, maybe that's the data structure @sechkova was thinking of, it seems to at least allow iterating etc.? :)

Another way of encapsulation would be to make them class attributes of the corresponding metadata classes. 

I'm really fine either way.

ivanayov commented:
I think enum covers all the cases, as @lukpueh also mentioned.

ivanayov commented:
I pushed an update with enum.

MVrachev commented on tests/repository_simulator.py:
```diff
@@ -368,9 +369,9 @@ def write(self):
+                f.write(self._fetch_metadata(RoleNames.ROOT.value, ver))
 
-        for role in ["timestamp", "snapshot", "targets"]:
+        for role in [RoleNames.TIMESTAMP.value, RoleNames.SNAPSHOT.value, RoleNames.TARGETS.value]:
```
You can make it shorter:
```suggestion
        for role in RoleNames:
```

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -61,14 +62,25 @@
 )
 
+
+class RoleNames(Enum):
+    ROOT: str = "root"
+    SNAPSHOT: str = "snapshot"
+    TARGETS: str = "targets"
+    TIMESTAMP: str = "timestamp"
+
+    @staticmethod
+    def all() -> set:
+        return set(map(lambda r: r.value, RoleNames))
```
```suggestion
class RoleNames(str, Enum):
    ROOT: str = "root"
    SNAPSHOT: str = "snapshot"
    TARGETS: str = "targets"
    TIMESTAMP: str = "timestamp"

    @staticmethod
    def all() -> set:
        return set(map(lambda r: r.value, RoleNames))
```

If you inherit the string class as well you will be able to do comparisons like:
```
if RoleNames.TARGETS == role
```
instead of doing:
```
if RoleNames.TARGETS.value == role
```
Other such examples from python docs:
https://docs.python.org/3/library/enum.html#others

MVrachev replied:
Additionally, when do we use `RoleNames.all()`? Why do we need it?

ivanayov replied:
It refers to @sechkova's comment for removing `TOP_LEVEL_ROLE_NAMES`. Enums in python don't have any not-breaking-the-rules option to access all their values in a list/set/etc.

ivanayov replied:
I'm fine to go with simple constants if that's the agreement, even though in perspective it doesn't feel like the best option. For example, one would need to always use 
```
from tuf.api.metadata import  Root, ROOT, Targets, TARGETS, Snapshot, SNAPSHOT, Timestamp, TIMESTAMP
```
instead of
```
from tuf.api.metadata import  Root, Targets, Snapshot, Timestamp, RoleNames
```

and if for some reason in future let's say timestamp for example is used in another context as well within TUF, using only TIMESTAMP constant might be confusing.

I changed to enum to address @sechkova's proposal, but if the agreement is to use constants, then we can keep `TOP_LEVEL_ROLE_NAMES `.

jku replied:
> For example, one would need to always use
> 
> from tuf.api.metadata import  Root, ROOT, Targets, TARGETS, Snapshot, SNAPSHOT, Timestamp, TIMESTAMP

This is why I've never done this*... it seems very hard to provide this as API that is not a repetitive mess but still offers the role names in a way that allows elegant code.

 *) I'm very happy someone is trying though: there could be a neat solution we haven't seen yet

MVrachev reviewed (changes requested):
Honestly, I am still uncertain about the usage of an enum here compared to using separate constants, and given that they will be uppercase I don't think they will be mistaken with something else.

Still, if we decide to use enums I have two suggestions:
1. when defining RoleNames make sure you inherit `str`, this will shorten the comparisons as I explained in my comment 
2. when iterating all of the roles make sure to do: `for role in RoleNames` instead of mentioning each of the roles explicitly.
I left a comment about that too.

**_EDIT_**: I think you can directly just inherit `StrEnum` with the same success as inheriting `Enum` and `str` separatly. 
For more info: https://docs.python.org/3/library/enum.html#others

jku reviewed (commented):
I don't have opinions on the actual data structure but I have a couple on the usage:
* a few legacy test files are modified: The split between legacy and metadata API is confusing so not your fault, but let's not start mixing the two implementations. Basically if the file did not import `tuf.api.metadata` already, it's probably legacy and we don't want to modify the file. so test_repository_tool, test_sig, test_updater should not be touched
* It's great that we try to avoid typos by avoiding hard-coded strings... but it should not come at at the expense of readability: I think going from `"root"` to `RoleNames.ROOT.value` is not a clear cut win as the latter is more than 3 times as long (!) and looks so complex -- I could not guess from the name that I'm looking at a simple role name.

jku commented:
It seems like the major reason to use an enum or a class is to avoid top level variable name pollution (so to avoid `from tuf.api.metadata import Root, ROOT, Timestamp, TIMESTAMP, ....`), correct?

What if instead of a new class we used the classes we are already importing? Root and others already have a class variable `_signed_type` which is already exposed as an instance property `type`... Maybe we could just simplify that so the class variable is called `type` (and is public): then we could just refer to `Root.type` to get the string without instantiating a Root?

(for reference: the class variable _signed_type is probably currently private to make it very clear it should not be modified, but as python does not have "class properties" that means we can't keep using a property in this case(?). I think solving the problem described in this issue is more useful than keeping the type private)

lukpueh commented:
> What if instead of a new class we used the classes we are already importing? Root and others already have a class variable `_signed_type` which is already exposed as an instance property `type`... Maybe we could just simplify that so the class variable is called `type` (and is public): then we could just refer to `Root.type` to get the string without instantiating a Root?

This is what I meant above with ...

> Another way of encapsulation would be to make them class attributes of the corresponding metadata classes.

One downside would be that we'd have to define a global `TOP_LEVEL_ROLE_NAMES` that uses these class variables after the class definitions, i.e. at the bottom of the metadata module. But I guess that's okay.

jku commented:
> One downside would be that we'd have to define a global `TOP_LEVEL_ROLE_NAMES` that uses these class variables after the class definitions, i.e. at the bottom of the metadata module.

I'm also not going to lose sleep over a single instance of duplicating the strings within the API implementation if that makes the API definition look neater: what's important is removing the magic strings from the code that imports `tuf.api.metadata`

Commit:
Create constants for top-level rolenames

This is a change in the metadata API to remove hardcoded rolenames
and use constants instead.
Fixes #1648

Signed-off-by: Ivana Atanasova <iyovcheva@iyovcheva-a02.vmware.com>

Commit:
Apply top-level rolenames constants in tests

This applies the use of constants of top-level rolenames in the
tests instead of the previously hardcoded strings.
Fixes #1648

Signed-off-by: Ivana Atanasova <iyovcheva@iyovcheva-a02.vmware.com>

ivanayov commented:
I've put the constants within the role classes. Agree that it's more clear, simple and convenient.

As those are both required private (we don't want to be able to change them, whatever "be able" means in python for private properties :D ), but at the same time we need them for static reference within Updater, and we don't want to break the file format which uses `_type`, and don't want to do a breaking API change for such a simple reason with changing `type` to `TYPE`, we came to and agreement with @jku that changing `_signed_type` to `type` and keeping the `_type` property would be the best possible option.
Yep, it's still not very nice to reference the public variable from the private property, but that's the least painful and best to use change, unless someone comes with a genius idea that we cannot see yet.

jku commented on tests/test_trusted_metadata_set.py:
```diff
@@ -15,6 +15,7 @@
 from tests import utils
 from tuf import exceptions
 from tuf.api.metadata import (
+    T,
```
is this needed?

jku commented on tests/test_updater_top_level_update.py:
```diff
@@ -35,7 +35,7 @@ class TestRefresh(unittest.TestCase):
         self.temp_dir = tempfile.TemporaryDirectory()
         self.metadata_dir = os.path.join(self.temp_dir.name, "metadata")
-        self.targets_dir = os.path.join(self.temp_dir.name, "targets")
+        self.targets_dir = os.path.join(self.temp_dir.name, Targets.type)
```
This is an unrelated "targets" and should not use Targets.type -- in fact it could be renamed "downloads" to be clear about it (it's the directory that client downloads the actual target files to)

jku commented on tuf/api/metadata.py:
```diff
@@ -61,14 +61,19 @@
     SignedSerializer,
 )
 
+_ROOT: str = "root"
+_SNAPSHOT: str = "snapshot"
+_TARGETS: str = "targets"
+_TIMESTAMP: str = "timestamp"
```
I can see why you added these, re-ordering the code wouldn't be enough to avoid them...

I might have just hard coded the strings in the three places inside this file (to make the code in those places easier to read) but I guess this works, it's just one more indirection.

That said, these variables should not need type annotations, right? I'd like to not include annotations unless they are necessary

jku reviewed (approved):
LGTM, left a couple of comments but I think none were blockers: please take a look and adapt or don't (leave a comment if you don't), either way works

jku commented:
> it's still not very nice to reference the public variable from the private property, but that's the least painful and best to use change, unless someone comes with a genius idea that we cannot see yet.

I think the genius idea is to just remove `_type` property: it's from the time when the API matching the file format was considered a bit more important than it is now. IMO the `_type` property is not worth the confusion it creates . But we can handle that in later PRs.

lukpueh commented:
> just remove _type property

Agreed!

>  it's from the time when the API matching the file format was considered a bit more important than it is now

I  actually suspect that in this case the (Python) implementation influenced the naming decision in the spec, to avoid shadowing the `type` global on json-to-dict deserialisation.

Regardless of this specific case, I still think that it very important that the spec file format is recognisable in the **reference** implementation.

ivanayov commented:
Thanks @jku. Just addressed your comments.
